sdk/core/System.BinaryData/src/BinaryData.cs

```diff
@@ -0,0 +1,260 @@
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by wrapping the passed in
+        /// <see cref="ReadOnlyMemory{Byte}"/>.
+        /// </summary>
+        /// <param name="data"></param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData FromBytes(ReadOnlyMemory<byte> data) =>
+            new BinaryData(data);
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by wrapping the passed in
+        /// byte array.
+        /// </summary>
+        /// <param name="data"></param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData FromBytes(byte[] data) =>
+            new BinaryData(data);
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance using the passed in string.
+        /// </summary>
+        /// <param name="data"></param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData FromString(string data) =>
+            new BinaryData(data);
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from the specified stream.
+        /// The passed in stream is not disposed by this method.
+        /// </summary>
+        /// <param name="stream">Stream containing the data.</param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData FromStream(Stream stream) =>
+            FromStreamAsync(stream, false).EnsureCompleted();
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from the specified stream.
+        /// The passed in stream is not disposed by this method.
+        /// </summary>
+        /// <param name="stream">Stream containing the data.</param>
+        /// <param name="cancellationToken">An optional<see cref="CancellationToken"/> instance to signal
+        /// the request to cancel the operation.</param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static async Task<BinaryData> FromStreamAsync(
+            Stream stream,
+            CancellationToken cancellationToken = default) =>
+            await FromStreamAsync(stream, true, cancellationToken).ConfigureAwait(false);
+
+        private static async Task<BinaryData> FromStreamAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken = default)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+            if (stream.CanSeek && stream.Length > int.MaxValue)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(stream),
+                    "Stream length must be less than Int32.MaxValue");
+            }
+            using var memoryStream = new MemoryStream();
```

Comment:
You have the stream length. You should initialize the memory stream to this length. otherwise it will grow, which means allocations and copies.

Reply:
The length is only available sometimes... streams that can be read, but not seeked, won't necessarily know how long they are.

Reply:
We only have the length if the stream is seekable.

Reply:
so, if we have a stream .length, .... :-)